Add transverse kicker

cemitch99 opened the pull request:
Added a thin transverse kicker element, corresponding to MAD-X elements 'hkicker', 'vkicker', 'kicker', and 'tkicker'.

- [x] Add element
- [x] Add Python equivalent
- [x] Add example
- [x] Add documentation

cemitch99 committed:
Examples for 3D space charge benchmarking

- Modified the initial beam size in the IOTA lens benchmark example.
- Added 2 benchmarks of 3D space charge for initial testing.
- Add documentation for 2 benchmarks with space charge.
- Add a benchmark example with space charge and periodic s-dependent focusing.
- Added an s-dependent example using a Kurth beam without space charge.
- Modified tolerance for IOTA lens benchmark example.
  Reduced tolerance to account for smaller initial beam size and
  improved preservation of invariants of motion.
- Modified tolerances of space charge examples to allow CI tests to
  pass when space charge is not active.

- Modified tolerance for space charge examples.
  These should fail unless space charge is turned on.

cemitch99 committed:
Update input_kurth_10nC.in

Selected numerical values for amr.n_cell, lattice.nslice, and geometry.prob_relative.

cemitch99 committed:
Add a thin transverse kicker.

cemitch99 committed:
Delete input_kurth_10nC.in

This is not part of this PR.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Add element docs.

cemitch99 committed:
Fix initial errors.

cemitch99 committed:
Correct Python element.

ax3l commented on src/particles/elements/Kicker.H:
```diff
@@ -0,0 +1,117 @@
+            amrex::ParticleReal dpx = m_xkick;
+            amrex::ParticleReal dpy = m_ykick;
+            if (m_unit == 1) {
+                  dpx = m_xkick / refpart.rigidity_Tm();
+                  dpy = m_ykick / refpart.rigidity_Tm();
```
Since the kick is already assigned:
```suggestion
                  dpx /= refpart.rigidity_Tm();
                  dpy /= refpart.rigidity_Tm();
```

ax3l commented on src/python/elements.cpp:
```diff
@@ -182,6 +182,16 @@ void init_elements(py::module& m)
+                amrex::ParticleReal const,
+                int const>(),
+             py::arg("xkick"), py::arg("ykick"), py::arg("units") = 0,
+             "A thin transverse kicker element"
```
```suggestion
             "A thin transverse kicker element. Kicks are for unit 0 dimensionless, or for 1 in T-m."
```

ax3l replied:
We probably can make this an enum once we have the example in place :)

ax3l commented on src/particles/elements/Kicker.H:
```diff
@@ -0,0 +1,117 @@
+
+} // namespace impactx
+
+#endif // IMPACTX_MULTIPOLE_H
```
```suggestion
#endif // IMPACTX_KICKER_H
```

ax3l commented on src/particles/elements/Kicker.H:
```diff
@@ -0,0 +1,117 @@
+    private:
+        amrex::ParticleReal m_xkick; //! horizontal kick strength
+        amrex::ParticleReal m_ykick; //! vertical kick strength
+        int m_unit; //! multipole index
```
```suggestion
        int m_unit; //! Kicks are for 0 dimensionless, or for 1 in T-m."
```

ax3l commented on src/particles/elements/Kicker.H:
```diff
@@ -0,0 +1,117 @@
+        /** Push all particles */
+        using BeamOptic::operator();
+
+        /** This is a multipole functor, so that a variable of this type can be used like a
+         *  multipole function.
```
```suggestion
        /** This is a transverse kicker functor, so that a variable of this type can be
         * used like a function.
```

cemitch99 committed:
Apply suggestions from code review

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

cemitch99 committed:
Add kicker example.

cemitch99 committed:
Add example doc.

cemitch99 committed:
Update CMakeLists.txt

cemitch99 committed:
Update README.rst

cemitch99 committed:
Update analysis_kicker.py

Relax tolerance slightly.

ax3l committed:
Kicker Update API: Enums & Strings

Update the C++ API to use an enum for the unit system and
the Python API & inputs file syntax to accept a string. That
makes the parameters at the call sites self-describing.

Also clarifies in doc strings what dimensionless is referring
to (units of magnetic rigidity of the reference particle).

ax3l reviewed (approved):
Thank you, that looks great.

I added a small commit using self-describing arguments at the call site for the kicker: a C++ enum and strings when describing the unit system of the Kicker. I also clarified a few details on the normalization of the dimensionless option.

cemitch99 reviewed (commented):
These changes look reasonable to me, thanks.  I agree that this should make the choice of units more transparent to the user.
